Update ADR 023 package versioning guidelines

aaronc opened the pull request:
ref: #6822

Given the discussions in #6905 which proposes moving everything to `beta`, I think it makes sense have some clear guidelines to help us make decisions about marking things as unstable vs stable.

This is my attempt at writing out some guidelines that specify clear criteria for marking things as `alpha`, `beta`, `v1`, `v2`, etc. Let's discuss and see if we can come to a point of alignment.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

amaurymartiny commented on docs/architecture/adr-023-protobuf-naming.md:
```diff
@@ -99,13 +99,67 @@ the best choices for the future.
+The following guidelines are recommended for marking packages as alpha or beta:
+* marking something as `alpha` or `beta` should be a last resort and just putting something in the
+stable package (i.e. `v1` or `v2`) should be preferred
+* a package *should* be marked as `alpha` *if and only if* there are active discussions to remove
+or significantly alter the package in the near future
+* a package *should* be marked as `beta` *if and only if* there is an active discussion to 
+significantly refactor/rework the functionality in the near future but not remove it
```
As described here, these 2 seems pretty similar to me: I don't see clearly the difference between `significantly alter the package` and `significantly refactor/rework the functionality`. For example, I sometimes here people say "x/bank" should be revamped, should it go as alpha or beta?

We could offer the same guarantees as [Kubernetes API](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning):
- alpha: "The API may change in incompatible ways in a later software release without notice."
- beta: "The schema and/or semantics of objects may change in incompatible ways in a subsequent beta or stable release. When this happens, we will provide instructions for migrating to the next version."

clevinson replied:
What you describe is how I read the current version:

"a package *should* be marked as `alpha` *if and only if* there are active discussions to **remove or** significantly alter the package in the near future"

aaronc replied:
Is there an update you would suggest @amaurymartiny ? I based this off of what I read in Kubernetes btw.

amaurymartiny commented on docs/architecture/adr-023-protobuf-naming.md:
```diff
@@ -99,13 +99,67 @@ the best choices for the future.
+* a package *should* be marked as `beta` *if and only if* there is an active discussion to 
+significantly refactor/rework the functionality in the near future but not remove it
+* modules *can and should* have types in both stable (i.e. `v1` or `v2`) and unstable (`alpha` or `beta`) packages.
+Experimental functionality to existing stable packages could be added as `alpha` for instance.
```
Does this mean that we suffix Alpha/Alpha1 in experimental types in a stable package?
e.g. `type MsgSendAlpha1`

aaronc replied:
No... stable packages are break checked so that wouldn't work

amaurymartiny replied:
Ah, no, I meant adding a new type `MsgSendAlpha1` in a stable package. Surely this does not break stability.

My question is:
- are we allowed to add a new type `MsgSendAlpha1` in a stable package?
- if yes, are we allowed to remove it later on? (that would break compat)

aaronc replied:
No we can't remove types from stable packages. Buf only allows exclusions on a per package basis. I think that's sensible too. Let's just add alpha/beta types to alpha/beta packages.

amaurymartiny replied:
> Let's just add alpha/beta types to alpha/beta packages.

Yeah, agreed with that.

```suggestion
```

amaurymartiny commented on docs/architecture/adr-023-protobuf-naming.md:
```diff
@@ -99,13 +99,67 @@ the best choices for the future.
+`v1` if it is actively used in immutable smart contracts.
+
+An alternative to creating `v2` packages is creating `v2` messages and RPC methods. For instance say we found a better API
+for sending coins, then we could add `MsgSendV2` to the existing bank package without needing to change `MsgSend`
```
I'm not sure I like adding `V2` suffix to types. We might end up with:
1. `bank.v1`'s `MsgSend` type
2. `bank.v1`'s `MsgSendV2` type
3. `bank.v2`'s `MsgSend` type
4. `bank.v2`'s `MsgSendV2` type

which could be confusing (esp 2&3). I'm okay to add Alpha/Beta suffixes to types though.

For new stable types, maybe instead of adding `V2`, we could just come up with a more descriptive name?

aaronc replied:
Good point. Let's remove this part. Let's just keep version suffixes in the package. Modules can always support multiple versions

clevinson replied:
agree with this point.

amaurymartiny commented:
could you also add an entry to this file's changelog?

aaronc commented:
I think this was ready to be merged. I'm going to go ahead and do that in case anyone has any objections.

aaronc commented:
Actually, maybe there were a couple of TODOs. Nevermind... will try to address.

aaronc commented:
@amaurymartiny I addressed all comments except the one where I added a question. Can you take a look again and then we'll see if we can get this merged?
